### PR TITLE
Standalone cinder

### PR DESCRIPTION
**Pull request by aglitke**
This PR adds a new provisioner called standalone-cinder.  The provisioner works by creating a cinder volume and mapping that volume's connection details to a PV using a natively supported VolumeSource.  Currently iscsi and rbd volumes are supported.  See README.md for more details on the architecture and flow.

The status of the code is a working prototype and I plan to actively collaborate with the community on further improvements.

Please see Issue 317 for the initial discussions about adding this provisioner.
https://github.com/kubernetes-incubator/external-storage/issues/317

### CODE REVIEW

**Comment by aglitke** [comment]
@j-griffith @humblec @dims FYI

Also, I tried to follow convention when updating the vendor code by using 'glide up -v ; glide-vc --use-lock-file'.  Please advise if I should do this differently.  Thanks!

**Comment by rootfs** [comment]
@aglitke need to resolve the conflict

**Comment by rootfs** [comment]
gofmt and golint 
```
Verifying /home/travis/gopath/src/github.com/kubernetes-incubator/external-storage/repo-infra/verify/go-tools/verify-gofmt.sh
!!! 'gofmt -s' needs to be run on the following files: 
./openstack/standalone-cinder/mapper.go
./openstack/standalone-cinder/provisioner.go
./openstack/standalone-cinder/mapper-rbd.go
./openstack/standalone-cinder/mapper-iscsi.go
./openstack/standalone-cinder/openstack.go
FAILED   /home/travis/gopath/src/github.com/kubernetes-incubator/external-storage/repo-infra/verify/go-tools/verify-gofmt.sh	0s
Verifying /home/travis/gopath/src/github.com/kubernetes-incubator/external-storage/repo-infra/verify/go-tools/verify-golint.sh
openstack/standalone-cinder/mapper-iscsi.go:26:7:warning: don't use ALL_CAPS in Go names; use CamelCase (golint)
openstack/standalone-cinder/mapper-iscsi.go:26:7:warning: exported const ISCSI_TYPE should have comment or be unexported (golint)
openstack/standalone-cinder/mapper-iscsi.go:27:7:warning: don't use ALL_CAPS in Go names; use CamelCase (golint)
openstack/standalone-cinder/mapper-iscsi.go:27:7:warning: exported const INITIATOR_NAME should have comment or be unexported (golint)
openstack/standalone-cinder/mapper-iscsi.go:38:9:warning: if block ends with a return statement, so drop this else and outdent its block (golint)
openstack/standalone-cinder/mapper-iscsi.go:103:8:warning: if block ends with a return statement, so drop this else and outdent its block (golint)
openstack/standalone-cinder/mapper-rbd.go:28:7:warning: don't use ALL_CAPS in Go names; use CamelCase (golint)
openstack/standalone-cinder/mapper-rbd.go:28:7:warning: exported const RBD_TYPE should have comment or be unexported (golint)
openstack/standalone-cinder/mapper.go:36:2:warning: struct field VolumeId should be VolumeID (golint)
openstack/standalone-cinder/mapper.go:105:49:warning: func parameter volumeId should be volumeID (golint)
openstack/standalone-cinder/mapper.go:134:51:warning: func parameter volumeId should be volumeID (golint)
openstack/standalone-cinder/mapper.go:152:47:warning: func parameter volumeId should be volumeID (golint)
openstack/standalone-cinder/mapper.go:170:2:warning: struct field cinderVolumeId should be cinderVolumeID (golint)
openstack/standalone-cinder/mapper.go:199:1:warning: exported function BuildPV should have comment or be unexported (golint)
openstack/standalone-cinder/mapper.go:199:48:warning: func parameter volumeId should be volumeID (golint)
openstack/standalone-cinder/openstack.go:38:6:warning: exported type Config should have comment or be unexported (golint)
openstack/standalone-cinder/openstack.go:41:3:warning: struct field AuthUrl should be AuthURL (golint)
openstack/standalone-cinder/openstack.go:43:3:warning: struct field UserId should be UserID (golint)
openstack/standalone-cinder/openstack.go:45:3:warning: struct field TenantId should be TenantID (golint)
openstack/standalone-cinder/openstack.go:47:3:warning: struct field TrustId should be TrustID (golint)
openstack/standalone-cinder/openstack.go:48:3:warning: struct field DomainId should be DomainID (golint)
openstack/standalone-cinder/openstack.go:106:9:warning: if block ends with a return statement, so drop this else and outdent its block (golint)
openstack/standalone-cinder/openstack.go:187:9:warning: if block ends with a return statement, so drop this else and outdent its block (golint)
openstack/standalone-cinder/provisioner.go:39:2:warning: const cinderVolumeId should be cinderVolumeID (golint)
openstack/standalone-cinder/provisioner.go:88:2:warning: var volumeId should be volumeID (golint)
openstack/standalone-cinder/provisioner.go:139:2:warning: var volumeId should be volumeID (golint)
openstack/standalone-cinder/provisioner.go:188:2:warning: var prId should be prID (golint)
FAILED   /home/travis/gopath/src/github.com/kubernetes-incubator/external-storage/repo-infra/verify/go-tools/verify-golint.sh	0s
Verifying /home/travis/gopath/src/github.com/kubernetes-incubator/external-storage/repo-infra/verify/go-tools/verify-gometalinter.sh
openstack/standalone-cinder/mapper.go:169:1:warning: mapperContext is unused (deadcode)
openstack/standalone-cinder/openstack.go:144::warning: declaration of "err" shadows declaration at openstack/standalone-cinder/openstack.go:139 (vetshadow)
FAILED   /home/travis/gopath/src/github.com/kubernetes-incubator/external-storage/repo-infra/verify/go-tools/verify-gometalinter.sh	6s
```

**Comment by wongma7** [comment]
feel free to add directories to .golintignore.... imo menial task of fixing golint should not block merging of the PR, it is something we can fix later as we did with snapshots.

File: openstack/standalone-cinder/provisioner.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+	var config *rest.Config
+	var err error
+	if *master != "" || *kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags(*master, *kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	prId := string(uuid.NewUUID())
+	if *id != "" {
```
Comment (rootfs) [inline openstack/standalone-cinder/provisioner.go]: see https://github.com/kubernetes-incubator/external-storage/blob/master/ceph/rbd/cmd/rbd-provisioner/main.go#L67
```go
// By default, we use provisioner name as provisioner identity.
```

File: openstack/standalone-cinder/mapper-iscsi.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	if secretName != "" {
+		ret.ISCSI.SessionCHAPAuth = true
+		ret.ISCSI.SecretRef = &v1.LocalObjectReference{
+			Name: secretName,
+		}
+	}
+	return ret, nil
+}
+
+func (m *iscsiMapper) AuthSetup(ctx provisionCtx) error {
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper-iscsi.go]: iSCSI CHAP is only in 1.7+

File: openstack/standalone-cinder/provisioner.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		return nil, err
+	}
+
+	connection, err := connectCinderVolume(p, volumeId)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to connect volume: %v", err)
+		return nil, err
+	}
+
```
Comment (rootfs) [inline openstack/standalone-cinder/provisioner.go]: reserve the volume

File: openstack/standalone-cinder/provisioner.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+	ctx := deleteCtx{p, pv}
+	mapper, err := newVolumeMapperFromPV(ctx)
+	if err != nil {
+		return err
+	}
+
+	mapper.AuthTeardown(ctx)
+
```
Comment (rootfs) [inline openstack/standalone-cinder/provisioner.go]: unreserve the volume

File: openstack/standalone-cinder/provisioner.go (hunk 1 above)
Comment (aglitke) [reply openstack/standalone-cinder/provisioner.go]: Ok.  I'll change it to use the provisionerName

File: openstack/standalone-cinder/mapper-iscsi.go (hunk 2 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper-iscsi.go]: How should I handle the kubernetes version dependencies for this?

**Comment by rootfs** [comment]
also need to import blockstorage extensions github.com/gophercloud/gophercloud/openstack/extensions/volumeactions

File: openstack/standalone-cinder/mapper-rbd.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+
+func (m *rbdMapper) BuildPVSource(ctx provisionCtx) (*v1.PersistentVolumeSource, error) {
+	mons := getMonitors(ctx.connection)
+	if mons == nil {
+		return nil, errors.New("No monitors could be parsed from connection info")
+	}
+	splitName := strings.SplitN(ctx.connection.Data.Name, "/", 2)
+
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper-rbd.go]: validate len(splitName)

Comment (aglitke) [reply openstack/standalone-cinder/mapper-rbd.go]: Will do

**Comment by rootfs** [comment]
please also provide a deployment and docker file

File: openstack/standalone-cinder/mapper-iscsi.go (hunk 2 above)
Comment (rootfs) [reply openstack/standalone-cinder/mapper-iscsi.go]: server git version should help, see
https://github.com/kubernetes-incubator/external-storage/blob/master/lib/controller/controller.go#L370

File: openstack/standalone-cinder/provisioner.go (hunk 3 above)
Comment (aglitke) [reply openstack/standalone-cinder/provisioner.go]: I would like to add TODOs in the code for this.  We should also call terminate_connection to tear down iscsi lun mappings on the storage server for example.

Comment (rootfs) [reply openstack/standalone-cinder/provisioner.go]: sgtm

**Comment by aglitke** [comment]
Sorry guys.  A bit new to the github PR process and the code validation tools for this project.

**Comment by rootfs** [comment]
build error
```
make[1]: Entering directory `/home/travis/gopath/src/github.com/kubernetes-incubator/external-storage/ceph/cephfs'
CGO_ENABLED=0 GOOS=linux go build -a -ldflags '-extldflags "-static"' -o cephfs-provisioner cephfs-provisioner.go
# github.com/kubernetes-incubator/external-storage/vendor/golang.org/x/sys/unix
../../vendor/golang.org/x/sys/unix/zsyscall_linux_amd64.go:1491: Msync redeclared in this block
	previous declaration at ../../vendor/golang.org/x/sys/unix/zsyscall_linux_amd64.go:1459
make[1]: *** [build] Error 2
```

**Comment by aglitke** [comment]
To resolve that build error I think I am going to have to close this PR and start over.  I resolved some vendor dependencies using the github merge ui and I think that broke things.

**Comment by rootfs** [comment]
@aglitke add a Makefile in openstack/standalone-cinder and update top level Makefile (when vendor is updated)

**Comment by aglitke** [comment]
I think this is ready now?

File: openstack/standalone-cinder/Makefile [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+.PHONY: all provisoner
+
+all: provisioner
+
```
Comment (rootfs) [inline openstack/standalone-cinder/Makefile]: also add a container target

File: openstack/standalone-cinder/README.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+intention is to support two modes of operation: a conventional
+cinder deployment with keystone managing authentication and
+providing the service catalog, and a standalone configuration where
+cinder is accessed directly.
+
+Conventional cinder deployments can be used by supplying a clound
+config file identical to the one you would use to configure an
+openstack cloud provider.
+
+### Workflows
```
Comment (rootfs) [inline openstack/standalone-cinder/README.md]: update it with reserve and unreserve

File: openstack/standalone-cinder/mapper-iscsi.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	// Create a secret for the CHAP credentials
+	secretName := getChapSecretName(ctx)
+	if secretName == "" {
+		glog.Info("No CHAP authentication secret necessary")
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper-iscsi.go]: glog.V(3)

File: openstack/standalone-cinder/mapper-iscsi.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+			"node.session.auth.password": []byte(ctx.connection.Data.AuthPassword),
+		},
+	}
+	namespace := ctx.options.PVC.Namespace
+	_, err := ctx.p.client.CoreV1().Secrets(namespace).Create(secret)
+	if err != nil {
+		glog.Errorf("Failed to create chap secret: %v", err)
+		return err
+	}
+	glog.Infof("Secret %s created", secretName)
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper-iscsi.go]: V(3) or V(4)

File: openstack/standalone-cinder/mapper-iscsi.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+		Type: "kubernetes.io/iscsi-chap",
+		Data: map[string][]byte{
+			"node.session.auth.username": []byte(ctx.connection.Data.AuthUsername),
+			"node.session.auth.password": []byte(ctx.connection.Data.AuthPassword),
+		},
+	}
+	namespace := ctx.options.PVC.Namespace
+	_, err := ctx.p.client.CoreV1().Secrets(namespace).Create(secret)
+	if err != nil {
+		glog.Errorf("Failed to create chap secret: %v", err)
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper-iscsi.go]: add namespace in error message, e.g. `failed to create chap secret in namespace %s: %v`

File: openstack/standalone-cinder/mapper-iscsi.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	if ctx.pv.Spec.ISCSI.SecretRef == nil {
+		glog.Info("No associated secret to delete")
+		return nil
+	}
+
+	secretName := ctx.pv.Spec.ISCSI.SecretRef.Name
+	secretNamespace := ctx.pv.Spec.ClaimRef.Namespace
+	err := ctx.p.client.CoreV1().Secrets(secretNamespace).Delete(secretName, nil)
+	if err != nil {
+		glog.Errorf("Failed to remove secret: %s, %v", secretName, err)
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper-iscsi.go]: add namespace

File: openstack/standalone-cinder/mapper-iscsi.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+		return err
+	}
+	glog.Infof("Successfully deleted secret %s", secretName)
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper-iscsi.go]: V(3) or V(4)

File: openstack/standalone-cinder/mapper-iscsi.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+		return err
+	}
+	glog.Infof("Secret %s created", secretName)
+	return nil
+}
+
+func (m *iscsiMapper) AuthTeardown(ctx deleteCtx) error {
+	// Delete the CHAP credentials
+	if ctx.pv.Spec.ISCSI.SecretRef == nil {
+		glog.Info("No associated secret to delete")
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper-iscsi.go]: V(3) or V(4)

File: openstack/standalone-cinder/mapper.go [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+		Name:             name,
+		Size:             sizeGB,
+		VolumeType:       volType,
+		AvailabilityZone: availability,
+	}
+
+	vol, err := volumes_v2.Create(p.volumeService, &opts).Extract()
+
+	if err != nil {
+		glog.Errorf("Failed to create a %d GB volume: %v", sizeGB, err)
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper.go]: GiB

File: openstack/standalone-cinder/mapper.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+		return "", err
+	}
+
+	glog.Infof("Created volume %v in Availability Zone: %v", vol.ID, vol.AvailabilityZone)
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper.go]: V(2)

File: openstack/standalone-cinder/mapper.go [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+		IP:        "127.0.0.1",
+		Initiator: initiatorName,
+	}
+	var rcv rcvVolumeConnection
+	err := volumeactions.InitializeConnection(p.volumeService, volumeID, &opt).ExtractInto(&rcv)
+	if err != nil {
+		glog.Errorf("failed to initialize connection :%v", err)
+		return volumeConnection{}, err
+	}
+	glog.Infof("Received connection info: %v", rcv)
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper.go]: V(3)

File: openstack/standalone-cinder/mapper.go [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+type volumeMapper interface {
+	BuildPVSource(ctx provisionCtx) (*v1.PersistentVolumeSource, error)
+	AuthSetup(ctx provisionCtx) error
+	AuthTeardown(ctx deleteCtx) error
+}
+
+func newVolumeMapperFromConnection(conn volumeConnection) (volumeMapper, error) {
+	switch conn.DriverVolumeType {
+	default:
+		msg := fmt.Sprintf("Unsupported volume type: %s", conn.DriverVolumeType)
```
Comment (rootfs) [inline openstack/standalone-cinder/mapper.go]: nit: persistent volume type (and below too)

File: openstack/standalone-cinder/provisioner.go [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+		return &controller.IgnoredError{
+			Reason: "identity annotation on PV does not match ours",
+		}
+	}
+	// TODO when beta is removed, have to check kube version and pick v1/beta
+	// accordingly: maybe the controller lib should offer a function for that
+
+	volumeID, ok := pv.Annotations[cinderVolumeID]
+	if !ok {
+		return errors.New("cinder volume id annotation not found on PV")
```
Comment (rootfs) [inline openstack/standalone-cinder/provisioner.go]: add volumeID in the message

File: openstack/standalone-cinder/provisioner.go [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+		glog.Errorf("Failed to unreserve volume: %v", err)
+		return err
+	}
+
+	err = deleteCinderVolume(p, volumeID)
+	if err != nil {
+		return err
+	}
+
+	glog.Infof("Successfully deleted cinder volume %s", volumeID)
```
Comment (rootfs) [inline openstack/standalone-cinder/provisioner.go]: V(2)

**Comment by rootfs** [comment]
ideally i like to see just provisioner in main package and it imports from mapper package. We can live with this structure for now and refactor after we get more mappers (e.g. fc)

File: openstack/standalone-cinder/Makefile (hunk 6 above)
Comment (aglitke) [reply openstack/standalone-cinder/Makefile]: Done.

File: openstack/standalone-cinder/README.md (hunk 7 above)
Comment (aglitke) [reply openstack/standalone-cinder/README.md]: Updated.

File: openstack/standalone-cinder/mapper-iscsi.go (hunk 8 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper-iscsi.go]: done

File: openstack/standalone-cinder/mapper-iscsi.go (hunk 9 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper-iscsi.go]: done

File: openstack/standalone-cinder/mapper-iscsi.go (hunk 10 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper-iscsi.go]: done

File: openstack/standalone-cinder/mapper-iscsi.go (hunk 11 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper-iscsi.go]: done

File: openstack/standalone-cinder/mapper-iscsi.go (hunk 12 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper-iscsi.go]: done

File: openstack/standalone-cinder/mapper-iscsi.go (hunk 13 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper-iscsi.go]: done

File: openstack/standalone-cinder/mapper.go (hunk 14 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper.go]: done

File: openstack/standalone-cinder/mapper.go (hunk 15 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper.go]: done

File: openstack/standalone-cinder/mapper.go (hunk 16 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper.go]: done

File: openstack/standalone-cinder/mapper.go (hunk 17 above)
Comment (aglitke) [reply openstack/standalone-cinder/mapper.go]: done

File: openstack/standalone-cinder/provisioner.go (hunk 18 above)
Comment (aglitke) [reply openstack/standalone-cinder/provisioner.go]: Changed to `errors.New(cinderVolumeID + " annotation not found on PV")`

File: openstack/standalone-cinder/provisioner.go (hunk 19 above)
Comment (aglitke) [reply openstack/standalone-cinder/provisioner.go]: done

**Comment by aglitke** [comment]
I agree wrt to adding the mapper package.  If it's okay with everyone, I'd prefer to do that refactoring post merging of this PR as I have plenty of other improvements and additions planned for this provisioner.

**Comment by rootfs** [comment]
provisioner lgtm
@wongma7 @tsmetana @xing-yang  please review vendor and snapshot changes

**Comment by wongma7** [comment]
glide/vendor looks ok

**Comment by rootfs** [comment]
snapshot changes lgtm

**Comment by rootfs** [comment]
@wongma7 can you create a new area for this? thanks

File: snapshot/pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 20]
```diff
@@ -930,7 +930,7 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
 
 		portID := loadbalancer.VipPortID
 		newOpts := []string{lbSecGroup.ID}
-		updateOpts := neutronports.UpdateOpts{SecurityGroups: &newOpts}
+		updateOpts := neutronports.UpdateOpts{SecurityGroups: newOpts}
```
Comment (xing-yang) [inline snapshot/pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: According to the following code, this should be a pointer so the original code is correct.  Is the definition in the SDK different?
https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/ports/requests.go#L118
	SecurityGroups      *[]string      `json:"security_groups,omitempty"`

File: snapshot/pkg/cloudprovider/providers/openstack/openstack_routes.go [hunk 21]
```diff
@@ -125,7 +125,7 @@ func updateAllowedAddressPairs(network *gophercloud.ServiceClient, port *neutron
 	origPairs := port.AllowedAddressPairs // shallow copy
 
 	_, err := neutronports.Update(network, port.ID, neutronports.UpdateOpts{
-		AllowedAddressPairs: &newPairs,
+		AllowedAddressPairs: newPairs,
```
Comment (xing-yang) [inline snapshot/pkg/cloudprovider/providers/openstack/openstack_routes.go]: Same here.  According to the following code, this should be a pointer so the original code is correct.  Is the definition in the SDK different?
https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/ports/requests.go#L118
		AllowedAddressPairs *[]AddressPair `json:"allowed_address_pairs,omitempty"`

File: snapshot/pkg/cloudprovider/providers/openstack/openstack_routes.go [hunk 22]
```diff
@@ -134,7 +134,7 @@ func updateAllowedAddressPairs(network *gophercloud.ServiceClient, port *neutron
 	unwinder := func() {
 		glog.V(4).Info("Reverting allowed-address-pairs change to port ", port.ID)
 		_, err := neutronports.Update(network, port.ID, neutronports.UpdateOpts{
-			AllowedAddressPairs: &origPairs,
+			AllowedAddressPairs: origPairs,
```
Comment (xing-yang) [inline snapshot/pkg/cloudprovider/providers/openstack/openstack_routes.go]: Same here.

File: snapshot/pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 20 above)
Comment (j-griffith) [reply snapshot/pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Yeah, not sure why the load_balancer and other non-related snapshot items were touched anyway... bad merge with something somewhere along the line maybe?

Comment (aglitke) [reply snapshot/pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: When I ran the glide up command it pulled in a version of gophercloud where those two fields were not pointers.  I am really not sure how that happened.  glide up is voodoo.  I'll try to fix that now.

**Comment by aglitke** [comment]
I fixed the gophercloud/snapshot vendor issue by specifying today's gophercloud HEAD as the requested version in glide.yaml.  For some reason glide was choosing an older version implicitly.

File: openstack/standalone-cinder/README.md [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+cluster is deployed on Openstack.  By mapping cinder
+volume connection information to native volume sources,
+persistent volumes can be attached and mounted using the
+standard kubelet machinery and without help from cinder.
+
+### Mapping cinder volumes to native PVs
+The provisioner works by mapping cinder volume connections 
+(iscsi, rbd, fc, etc) to the corresponding native/raw kubernetes
+volume types.  New cinder types can be supported in the provisioner
+by creating a new implementation of the volunmeMapper interface.  The
```
Comment (xing-yang) [inline openstack/standalone-cinder/README.md]: s/volunmeMapper/volumeMapper

File: openstack/standalone-cinder/README.md [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+
+### Connecting to cinder
+The provisioner directly uses the gophercloud SDK to connect to
+cinder (as opposed to use of the cloudprovider interface).  The
+intention is to support two modes of operation: a conventional
+cinder deployment with keystone managing authentication and
+providing the service catalog, and a standalone configuration where
+cinder is accessed directly.
+
+Conventional cinder deployments can be used by supplying a clound
```
Comment (xing-yang) [inline openstack/standalone-cinder/README.md]: s/clound/cloud

**Comment by wongma7** [comment]
merging: please fix typos in a follow-up! thanks all!

**Pull request merged**